internal/envoy/cors/cors.go

```diff
@@ -0,0 +1,94 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cors
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/davecgh/go-spew/spew"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_config_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+const (
+	HeaderOrigin                        = "Origin"
+	HeaderAccessControlAllowOrigin      = "Access-Control-Allow-Origin"
+	HeaderAccessControlAllowCredentials = "Access-Control-Allow-Credentials"
+	HeaderAccessControlMaxAge           = "Access-Control-Max-Age"
+)
+
+func ConfigureCORSOnRoute(corsPolicy *envoy_config_route_v3.CorsPolicy, route *envoy_config_route_v3.Route, origins []string, funcName string) {
+	if corsPolicy == nil {
+		return
+	}
+
+	// request headers to add for upstream
+	for _, origin := range origins {
+		route.RequestHeadersToAdd = append(route.RequestHeadersToAdd, &envoy_config_core_v3.HeaderValueOption{
+			Header: &envoy_config_core_v3.HeaderValue{
+				Key:   HeaderOrigin,
+				Value: origin,
+			},
+			Append: wrapperspb.Bool(true),
+		})
+	}
+
+	// response headers to add
+	// for _, origin := range origins {
+	route.ResponseHeadersToAdd = append(route.ResponseHeadersToAdd, &envoy_config_core_v3.HeaderValueOption{
+		Header: &envoy_config_core_v3.HeaderValue{
+			Key:   HeaderAccessControlAllowOrigin,
+			Value: strings.Join(origins, ","),
+			// Value: origin,
+		},
+		Append: wrapperspb.Bool(false),
+	})
+	// }
+
+	if corsPolicy.MaxAge != "" {
+		route.ResponseHeadersToAdd = append(route.ResponseHeadersToAdd, &envoy_config_core_v3.HeaderValueOption{
+			Header: &envoy_config_core_v3.HeaderValue{
+				Key:   HeaderAccessControlMaxAge,
+				Value: corsPolicy.MaxAge,
+			},
+			Append: wrapperspb.Bool(false),
+		})
+	}
+
+	if corsPolicy.AllowCredentials != nil && corsPolicy.AllowCredentials.Value {
+		route.ResponseHeadersToAdd = append(route.ResponseHeadersToAdd, &envoy_config_core_v3.HeaderValueOption{
+			Header: &envoy_config_core_v3.HeaderValue{
+				Key:   HeaderAccessControlAllowCredentials,
+				Value: "true",
+			},
+			Append: wrapperspb.Bool(false),
+		})
+	}
+
+	fmt.Printf("ConfigureCORSOnRoute.%s: origins=%v\n", funcName, origins)
```

Comment:
probably for future reference we should wrap this into logger

Reply:
I've changed it to use `logger` now.